Java/src/com/couchbase/litecore/fleece/MDict.java

```diff
@@ -13,92 +13,244 @@
+    public long count() {
+        return _count;
     }
 
-    public boolean set(String key, MValue val) {
-        return set(_handle, key, val._handle);
+    public FLSharedKeys getSharedKeys() {
+        return getContext().getSharedKeys();
     }
 
-    public boolean remove(String key) {
-        return remove(_handle, key);
-    }
+    /* Public methods */
 
     public boolean clear() {
-        return clear(_handle);
+        if (!isMutable())
+            throw new IllegalStateException("Cannot clear a non-mutable MDict");
+
+        if (_count == 0)
+            return true;
+
+        mutate();
+        _map.clear();
+
+        if (_dict != null) {
+            FLDictIterator itr = new FLDictIterator();
+            try {
+                itr.begin(_dict);
+                String key;
+                SharedKeys sk = new SharedKeys(getSharedKeys());
+                while ((key = SharedKeys.getKey(itr, sk)) != null) {
+                    _map.put(key, MValue.EMPTY);
+                    if (!itr.next())
+                        break;
+                }
+            } finally {
+                itr.free();
+            }
+        }
+
+        _count = 0;
+        return true;
     }
 
-    public void encodeTo(Encoder enc) {
-        encodeTo(_handle, enc._handle);
+    public boolean contains(String key) {
+        if (key == null)
+            throw new IllegalArgumentException("The key cannot be null.");
+
+        MValue v = _map.get(key);
+        if (v != null)
+            return !v.isEmpty();
+        else
+            return _dict != null && _dict.getSharedKey(key, getSharedKeys()) != null;
     }
 
-    @Override
-    protected void finalize() throws Throwable {
-        free();
-        super.finalize();
+    public MValue get(String key) {
+        if (key == null)
+            throw new IllegalArgumentException("The key cannot be null.");
+
+        MValue v = _map.get(key);
+        if (v == null) {
+            FLValue value = _dict != null ? _dict.getSharedKey(key, getSharedKeys()) : null;
+            v = value != null ? setInMap(key, new MValue(value)) : MValue.EMPTY;
+        }
+        return v;
     }
 
-    @Override
-    public void free() {
-        if (_handle != 0L && !_managed) {
-            free(_handle);
-            _handle = 0L;
+    public List<String> getKeys() {
+        List<String> keys = new ArrayList<>();
+        for (Map.Entry<String, MValue> entry : _map.entrySet()) {
+            MValue value = entry.getValue();
+            if (!value.isEmpty())
+                keys.add(entry.getKey());
         }
+
+        if (_dict != null) {
+            FLDictIterator itr = new FLDictIterator();
+            try {
+                itr.begin(_dict);
+                String key;
+                SharedKeys sk = new SharedKeys(getSharedKeys());
+                while ((key = SharedKeys.getKey(itr, sk)) != null) {
+                    if (!_map.containsKey(key))
```

Comment:
question: Does `getKeys()` return the key for its value is `MValue.EMPTY`?

Reply:
It shouldn't. Why do you think it will return the empty value keys?

Reply:
This code works as expected. Originally I misunderstood the behavior if the value is EMPTY. no issue here.